catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardImpl.java

```diff
@@ -156,10 +159,22 @@ public MetacardImpl(Metacard metacard, MetacardType type) {
             throw new IllegalArgumentException(
                     MetacardType.class.getName() + " instance should not be null.");
         }
-        map = new HashMap<>();
-        for (AttributeDescriptor attribute : metacard.getMetacardType().getAttributeDescriptors()) {
-            map.put(attribute.getName(), metacard.getAttribute(attribute.getName()));
+
+        if (metacard.getSourceId() != null) {
+            setSourceId(metacard.getSourceId());
         }
+
+        map = new ConcurrentHashMap<>();
+
+        metacard.getMetacardType()
+                .getAttributeDescriptors()
+                .stream()
+                .filter(attributeDescriptor -> attributeDescriptor != null)
+                .filter(attributeDescriptor -> attributeDescriptor.getName() != null)
+                .filter(attributeDescriptor -> metacard.getAttribute(attributeDescriptor.getName())
+                        != null)
+                .forEach(attributeDescriptor -> map.put(attributeDescriptor.getName(),
+                        metacard.getAttribute(attributeDescriptor.getName())));
```

Comment:
instead of a `.forEach`, why not do a collect and set whole container to the result?
eg
`map = ...
  .map(AttributeDescriptor::getName)
  .collect(Collectors.toMap(Function.identity(), metacard::getAttribute));